### PR TITLE
Move threads over to new schema migrations

### PR DESCRIPTION
**Pull request by nronas**
#### Summary
implements: https://mattermost.atlassian.net/browse/MM-33278

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-33278

#### Release Note

```release-note
* Moves threads gorp driven migrations to schema-based migrations
```

**NOTE** Relies on Posts migrations to be merged first

### CODE REVIEW

File: db/migrations/mysql/000023_create_threads.up.sql [hunk 1]
```diff
@@ -0,0 +1,43 @@
+CREATE TABLE IF NOT EXISTS Threads (
+    PostId varchar(26) NOT NULL,
+    ChannelId varchar(26) DEFAULT NULL,
```
Comment (mgdelacroix) [inline db/migrations/mysql/000023_create_threads.up.sql:3]: This one should be created outside the table, as it's part of the 529 migration

File: db/migrations/mysql/000023_create_threads.up.sql [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    ReplyCount bigint(20),
+    LastReplyAt bigint(20),
+    Participants text,
+    PRIMARY KEY (PostId)
+) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4;
+
+SET @preparedStatement = (SELECT IF(
+    (
+        SELECT COUNT(*) FROM INFORMATION_SCHEMA.COLUMNS
+        WHERE table_name = 'Reactions'
```
Comment (mgdelacroix) [inline db/migrations/mysql/000023_create_threads.up.sql:13]: As this alter is affecting Reactions, is this its PR? We're missing an alter for ChannelId, so this might be the original paste without modifications for that one...

**Review by mgdelacroix — CHANGES_REQUESTED** [review]

File: go.mod [hunk 3]
```diff
@@ -94,9 +95,8 @@ require (
 	github.com/pelletier/go-toml v1.9.1 // indirect
 	github.com/pierrec/lz4/v4 v4.1.7 // indirect
 	github.com/pkg/errors v0.9.1
-	github.com/prometheus/client_golang v1.10.0
+	github.com/prometheus/client_golang v1.11.0
```
Comment (agnivade) [inline go.mod:98]: All of these go.mod updates seem unrelated. Is that okay?

File: db/migrations/mysql/000031_create_threads.up.sql [hunk 4]
```diff
@@ -0,0 +1,42 @@
+CREATE TABLE IF NOT EXISTS Threads (
+    PostId varchar(26) NOT NULL,
+    ReplyCount bigint(20),
+    LastReplyAt bigint(20),
+    Participants text,
```
Comment (agnivade) [inline db/migrations/mysql/000031_create_threads.up.sql:5]: This column is now json. Are we incorporating the latest schema changes from master or doing that later?

File: db/migrations/mysql/000032_thread_memberships.down.sql [hunk 5]
```diff
@@ -0,0 +1 @@
+DROP TABLE IF EXISTS ThreadMemberships;
```
Comment (agnivade) [inline db/migrations/mysql/000032_thread_memberships.down.sql:1]: There is an inconsistency with the other down.sql. In the earlier one, we remove the indexes, columns and then the table. But here we are removing only the table. We should be consistent every where.

File: db/migrations/mysql/000031_create_threads.up.sql (hunk 4 above)
Comment (isacikgoz) [reply db/migrations/mysql/000031_create_threads.up.sql:5]: Actually the whole thing PRs currently pointing to the [MM-31622/schema-migrations](https://github.com/mattermost/mattermost-server/tree/MM-31622/schema-migrations) so this is being updated to that version. Once we cleanup and merge I'm planning to update to the recent changes and merge the feature branch.

File: go.mod (hunk 3 above)
Comment (isacikgoz) [reply go.mod:98]: Hmm, I didn't noticed that, it shouldn't include dependency changes.

File: db/migrations/mysql/000032_thread_memberships.down.sql (hunk 5 above)
Comment (isacikgoz) [reply db/migrations/mysql/000032_thread_memberships.down.sql:1]: Oops 👍

File: db/migrations/mysql/000031_create_threads.up.sql (hunk 4 above)
Comment (agnivade) [reply db/migrations/mysql/000031_create_threads.up.sql]: Sounds good.

**Review by agnivade — APPROVED** [review]

**Review by mgdelacroix — APPROVED** [review]
